Libraries/LibCore/EventLoop.cpp

```diff
@@ -409,10 +416,114 @@ void EventLoop::post_event(Object& receiver, NonnullOwnPtr<Event>&& event)
     m_queued_events.empend(receiver, move(event));
 }
 
+EventLoop::SignalHandlers::SignalHandlers(int signo)
+    : m_signo(signo)
+    , m_original_handler(signal(signo, EventLoop::handle_signal))
+{
+#ifdef EVENTLOOP_DEBUG
+    dbg() << "Core::EventLoop: Registered handler for signal " << m_signo;
+#endif
+}
+
+EventLoop::SignalHandlers::~SignalHandlers()
+{
+    if (m_valid) {
+#ifdef EVENTLOOP_DEBUG
+        dbg() << "Core::EventLoop: Unregistering handler for signal " << m_signo;
+#endif
+        signal(m_signo, m_original_handler);
+    }
+}
+
+void EventLoop::SignalHandlers::dispatch()
+{
+    for (auto& handler : m_handlers)
+        handler.value(m_signo);
+}
+
+int EventLoop::SignalHandlers::add(Function<void(int)>&& handler)
+{
+    int id = ++EventLoop::s_next_signal_id; // TODO: worry about wrapping and duplicates?
+    m_handlers.set(id, move(handler));
+    return id;
+}
+
+bool EventLoop::SignalHandlers::remove(int handler_id)
+{
+    ASSERT(handler_id != 0);
+    return m_handlers.remove(handler_id);
+}
+
+void EventLoop::dispatch_signal(int signo)
+{
+    // We need to protect the handler from being removed while handling it
+    TemporaryChange change(s_handling_signal, signo);
+    auto handlers = s_signal_handlers.find(signo);
+    if (handlers != s_signal_handlers.end()) {
+#ifdef EVENTLOOP_DEBUG
+        dbg() << "Core::EventLoop: dispatching signal " << signo;
+#endif
+        handlers->value.dispatch();
+    }
+}
+
+void EventLoop::handle_signal(int signo)
+{
+    ASSERT(signo != 0);
+    // We MUST check if the current pid still matches, because there
+    // is a window between fork() and exec() where a signal delivered
+    // to our fork could be inadvertedly routed to the parent process!
```

Comment:
What does this mean? Signals are sent to a PID, and the fork will have a different PID?
So how would a signal delivered to the child end up in the parent?

Reply:
Because the pipe doesn't get closed until `exec`. `fork` inherits file descriptors. So, if you fork, if anything sends a signal to that new PID and that signal handler gets executed before `exec`, the signal handler will write to the pipe, and the parent process reads from the other end thinking it needs to asynchronously handle that signal.

Reply:
By the way, it's not like `fork()` is always soon followed by `exec()`. It's perfectly valid to `fork()` without `exec`ing, and any design for handling signals in the `EventLoop` should take that into account.

Reply:
This is true on Linux, too. A few months ago I spent quite some time at work tracking down a bug where this is what ended up happening. Because the `exec` (on Linux) took a long time because it was a large program and the system was relatively busy just at the same time, the parent process sent a `SIGTERM` thinking it must be not responding, but the signal handler in the fork sent it into the cloned file descriptor, so the parent ended up receiving it in its own pipe shutting itself down by accident.

Reply:
Ah yes, that's awkward. Perhaps we should solve this with a custom fd flag a la `O_CLOEXEC`.
`fcntl(fd, FD_SETFL, ... | FD_NOINHERIT);` or something like that.

Reply:
Yeah that would do it, too. This is the "portable" solution ;-) Not that we really aim to be...

Reply:
Actually, the problem I see with something like `FD_NOINHERIT` is that there is no way to detect that the process had been forked, unless you compare the PID. The only thing you would get is an error trying to write to the now closed file descriptor, but what if that file descriptor was re-allocated to something completely different already, like a socket? Seems like this is the only reliable way to solve this problem.

Reply:
We've been thinking about introducing signalfd, and perhaps it would let each process read its own signals.

Reply:
We can always change it once that's in place. The small overhead really isn't a big deal.